### PR TITLE
DBDAART-4270_Separate_Post_Job_Statistics_Runs

### PR DESCRIPTION
**Pull request by FICU-tensileai**
## What is this and why are we doing it?
Currently, the Post Job Statistics and Metadata runs in the notebook are tied to the job run itself. This code separates the job code by adding a parameter into the Runner called "run_stats_only" which defaults to False (which means no notebook changes are needed for current job runs.) If you'd like to run the notebook for statistics then change this to True and it will skip all inserts into the permanent tables and run the stats on the da_run_id and state_codes you specify

* Link to the Jira ticket for this change: https://jiraent.cms.gov/browse/DBDAART-4270

## What are the security implications from this change?
Nothing has changed with regards to who can run this code. The same permissions for the job and runners remain intact.

## How did I test this?
Ran the notebook jobs with run_stats_only = True on the runners.

## Should there be new or updated documentation for this change? (Be specific.)
Yes, once we get official documentation, this change should be added.

## PR Checklist
- [x] The JIRA ticket number and a short description is in the subject line
- [x] My code follows any applicable [style guides](https://cms-dataconnect.atlassian.net/wiki/search?text=style%20guide)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: taf/APL/APL_Runner.py [hunk 1]
```diff
@@ -14,14 +14,16 @@ def __init__(self,
...
+                 file_version: str,
+                 run_stats_only: bool = False):
         super().__init__(da_schema,
                          reporting_period,
                          state_code,
                          run_id,
                          job_id,
                          file_version)
 
+        self.run_stats_only = run_stats_only
```
Comment (RIE9-adhoc) [inline taf/APL/APL_Runner.py:26]: The notebook is passing run_stats_only as a string, which is always resolving to True later on when doing `if run_stats_only`. I believe this variable needs to be explicitly casted to a bool everywhere.

Comment (FICU-tensileai) [reply taf/APL/APL_Runner.py:26]: Hmmm. I have the code here with run_stats_only: bool = False. Are you sure you didn't look at an older checkin? I had to make a couple of updates for the reason you stated.

Comment (FICU-tensileai) [reply taf/APL/APL_Runner.py:26]: BTW: You can't even cast.

Comment (RIE9-adhoc) [reply taf/APL/APL_Runner.py:26]:
```
>>> bool(int('0'))
False
>>> bool(int('1'))
True
```
This might work better if we specify 0/1 in the parameter.

**Comment by FICU-tensileai** [comment]
I think the change I put is safe though. It’ll take True or 1. Otherwise it forces to False. This way we don’t accidently run it wit a mis-type.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: taf/APL/APL_Runner.py [hunk 2]
```diff
@@ -14,14 +14,16 @@ def __init__(self,
                  state_code: str,
                  run_id: str,
                  job_id: int,
-                 file_version: str):
+                 file_version: str,
+                 run_stats_only: bool = False):
```
Comment (RIE9-adhoc) [inline taf/APL/APL_Runner.py:18]: I believe APR and DE are missing.

**Comment by FICU-tensileai** [comment]
They are. They didn't fit the "build()" pattern I was using for stopping inserts. Looks like some inserts are outside of that pattern.

________________________________

**Comment by RIE9-adhoc** [comment]
Should those be refactored then? Having these 2 file types included in having metadata running independently would make the code more consistent.

**Comment by FICU-tensileai** [comment]
I definitely agree with that. Not sure why the build pattern wasn't followed. Judging by the code creation it could have been one of the first classes converted before build()
________________________________

File: taf/APL/APL_Runner.py (hunk 2 above)
Comment (FICU-tensileai) [reply taf/APL/APL_Runner.py:18]: Ok, updates are checked-in.

File: taf/DE/DE0001BASE.py [hunk 3]
```diff
@@ -687,7 +687,7 @@ def create_base(self, tblname):
 
         z += f""",coalesce(b.MDCD_ENRLMT_DAYS_YR,0) as MDCD_ENRLMT_DAYS_YR
                  ,coalesce(b.CHIP_ENRLMT_DAYS_YR,0) as CHIP_ENRLMT_DAYS_YR
-                 ,coalesce(b.EL_DTS_SPLMTL,0) as EL_DTS_SPLMTL
+                 ,coalesce(b.EL_DTS_SPLMTL,0) as EL_self.run_stats_only = self.__forceBool__(run_stats_only)_SPLMTL
```
Comment (RIE9-adhoc) [inline taf/DE/DE0001BASE.py:690]: I think this was changed unintentionally.

File: taf/TAF_Runner.py [hunk 4]
```diff
@@ -1057,6 +1057,15 @@ def repack_code(self, code: str, str_num: str, padzero: bool = False, num_length
 
             return repack
 
+    # Private function to make sure runtime param is forced to boolean.
+    # Python evaluates everything to True except 0 and True
+    # so we're forcing it to False otherwise for safety reasons
+    def __forceBool__(self, _boolean_: int = 0):
+        if str(_boolean_) not in ('0', '1'):
+            self.logger.warning(f"'run_stats_only' parameter is passed as '{_boolean_}' but converted to False. If you meant 'True' please pass '1'")
```
Comment (RIE9-adhoc) [inline taf/TAF_Runner.py:1065]: I'd opt for a raised exception if the parameter is invalid and keep it to "Please pass 0 or 1". I think this statement would be easy to ignore in the event of an unexpected value since execution would continue.

File: taf/OT/OT.py [hunk 5]
```diff
@@ -44,7 +44,7 @@ def AWS_Extract_Line(self, TMSIS_SCHEMA, DA_SCHEMA, fl2, fl, tab_no, _2x_segment
         # Create a temporary line file
         z = f"""
             create or replace temporary view {fl2}_LINE_IN as
-            select
+            cache select
```
Comment (RIE9-adhoc) [inline taf/OT/OT.py:47]: Was this change intentional?

File: taf/TAF_Runner.py [hunk 6]
```diff
@@ -1016,6 +1015,7 @@ def __create_formated_lkup(self, filetyp: str, fl: str, fl2: str, rpt_out: str,
 
             rstr = hash(time.time())
             df_sp.createOrReplaceTempView(f"pgm_audt_cnt_lkp_{rstr}")
+            print(df_sp)
```
Comment (RIE9-adhoc) [inline taf/TAF_Runner.py:1018]: Was this a debugging statement? Maybe it could be removed or changed to logger.debug?

File: taf/DE/DE0001BASE.py (hunk 3 above)
Comment (FICU-tensileai) [reply taf/DE/DE0001BASE.py:690]: Yes. Good catch.

File: taf/OT/OT.py (hunk 5 above)
Comment (FICU-tensileai) [reply taf/OT/OT.py:47]: Nope. Was working on something locally and that snuck in. Thanks.

File: taf/TAF_Runner.py (hunk 6 above)
Comment (FICU-tensileai) [reply taf/TAF_Runner.py:1018]: Not needed. Debugging locally can make a mess of things. Supposedly, there is a way to step debug for remote Databricks but haven't figured it out yet.

File: taf/TAF_Runner.py (hunk 4 above)
Comment (FICU-tensileai) [reply taf/TAF_Runner.py:1065]: Fair enough.

**Comment by RIE9-adhoc** [comment]
@FICU-tensileai Code looks good. Could you paste a link to some runs using this code and the extra option?

**Comment by FICU-tensileai** [comment]
This one shows the ValueError
 [Run of TAF_Runner_Long_Term_Care_Claims of TAF_Runner_Long_Term_Care_Claims_Post_Job - Databricks](https://cms-dataconnect-val.cloud.databricks.com/?o=955724715920583#job/517364346938826/run/818065543960554)

This one shows the runner.init() info message skipping inserts and running post job only
[Run of TAF_Runner_Outpatient_Claims of TAF_Runner_Outpatient_Claims_Post_Job - Databricks](https://cms-dataconnect-val.cloud.databricks.com/?o=955724715920583#job/965362078910289/run/586891447650141)

**Review by RIE9-adhoc — APPROVED** [review]
